### ISSUE TITLE
Work-in-progress article for Thema, ECSS, and GPS

### ISSUE THREAD

**Issue by badlydrawnrob** [badlydrawnrob/thema-css#1]
# The Tao of stylesheets

- This is really long and needs to be split out into separate mini-articles or FAQs
- Would Thema stand up as the main way to do things? (using `#page` ID as a scope)
- Does ECSS help us to quickly see where the global styles are?

> Or, how to stay sane with CSS and HTML.
> These are my current thoughts, but I may change my mind!
> Time will tell how I use GPS, and if Thema Styles sticks around.

I’m currently trialling out **GPS**[^1] methodology: it has a simpler form than **[BEM](https://css-tricks.com/bem-101/)**, **[ECSS](https://benfrain.com/fun-css-naming-convention-explained/)**, or **[Tailwind](https://tailwindcss.com/)**, and is interested in reducing complexity and repetition while maintaining isolation and inheritance-proof styles. ECSS asks you to write a unique `.scoped-Class` for each and every item (similar to BEM); heavily relying on isolation to avoid inheritance problems. The biggest problem with these frameworks is:

- **Complexity.** Too many rules to follow.
- **Ugliness.** Does your HTML look like tag soup?
- **Verbosity.** Too design-heavy. Not simple enough.
- **Bloat.** To repetitive, leading to lots and lots of code.

GPS advocates for code reduction, less repetition, shared styles, and unique page styles. It’s a good theory, but we must be careful with inheritance — not accidentally changing styles, or adding in complexity that needn’t be there. GPS stands for _Global_, _Page_, _Section_ (which we’ll come to shortly). I’m sticking with some ECSS naming conventions.

## It starts with a specimen file

> Where plain HTML can be styled, let it be styled there first.
> Aim to make it accessible, especially colours and typography.
> Then, and only then, should you reach for GPS.

Your stylesheets should begin by creating a specimen file to visualise the lion’s share of the base-level HTML elements. This includes `normalize.css` for resets; typography and colour that’s accessible; block-level elements such as tables, inline code, links, etc; margins and padding. Your `specimen.html` file should preview your baseline-grid if you have one, vertical rhythm, and base-level elements that are widely used. Your specimen file should contain all the essentials and be pleasing to the eye.

Try to think about the essence of your sites look and feel upfront. Don’t overplan, but knowing how a Markdown file might compile, and how different combinations of elements fit together is a good start — can you combine elements and move them around without breaking the styling? That’s important. It gives you a really tight and solid base to work from. If you’ve got a design-system sketched out, you can decide how much of it can be made up from these simple base-level HTML building blocks, before you add in any global styles and classes. After that, you can add to your design-system one-by-one as you go. 

Consider also how your base-level elements are going to be viewed. For instance, in my [Anki flashcards](https://github.com/badlydrawnrob/anki), the `h1` is _always_ going to be wrapped in a `.gl-Card` class, within a global component. I might as well style the `font-*`, `margin`, and `padding` styles in my `/style/html/` folder _as if_ it’s within that card. This is debatable, but it’s a good idea to ask these questions before you start.

After you’ve done that, it’s on to the meat of the design-system and page-level styles.

## Global, page, section

> There’s many ways to write CSS, but here are some useful rules to follow.

We have three conceptual areas for our CSS: Globals, Pages, and Sections. Globals are your design-system, the styles likely to repeat multiple times throughout your app. Pages are the main wrapper for your unique views (or pages of your website), and sections are unique areas of content specific to that view, or page. The idea is that a piece of content can move up and down these conceptual areas, depending on how unique it is. Here’s some examples:

- A `.gl-Component`, such as a `.gl-Box`, or `.gl-PullQuote`.
  - These can have children, which we’ll come to later.
- Individual views or pages, such as an `#about` page. These are IDs.
  - You can add repeating `.item` classes scoped to the page.
  - Within the `#about` page, perhaps we have a section that’s called `#quotes`, which in turn could have it’s own children.

Remember, our `#quotes` section is _unique_ to the `#about` page. If it were to become apparent that our quotes were needed elsewhere in the page, we could migrate it from a section to a repeating class unique to that page.[^2] If the quotes become popular, and we’d like to add them to another part of our app, we can migrate it up to a `.gl-`obal element in our design-system. 

We aim to avoid scope leaks, meaning styles that can effect our main CSS outside of their intended use (for example, a `.quote` class should never affect our generic `blockquote`). We also try to avoid adding elements or class names if they’re not necessary — remove dead code and empty classes (a `.dead-class { }` we’ve created that has no styles).

## Globals

> For the time being, I’m sticking with ECSS `.gl-CamelCase` , but wherever possible using plain HTML elements to style, such as `.gl-Nav ul li`. The great thing about ECSS is that the class names are _very_ descriptive.

It’s important to be consistant. I’ve used this style for a good number of years, so heavily nesting styles (like GPS asks for) is a habit I’m not into. Every other style in the codebase is going to be `.snake-case` so this should visually separate the global styles in my design-system from page-specific ones.

### Start with the simplest thing possible (your design-system)

> **Be brutal.** See how far you can get with simple layouts, base-level elements, and a very simple design-system. CSS is a royal pain in the ass, and can be a time sink. Try using plain old HTML until you can’t go any further — turns out you can go pretty far!

I start with an attitude of "I don't want to build that". I try to go as far as I can with basic HTML before I reach for anything else.

Your `.gl-`obal styles should be provided _after_ your basic `specimen.html` file has been produced. Take care to notice anything that could be added to your raw HTML styles, without having to use `.gl-` classes. For example, rather than using a `.gl-Image-full-width` you can simply style your `<image>` elements as `img { max-width: 100%}` and you’re done!

We’re calling our Global styles a design-system, which basically means anything repeated across the website. It’s helpful to have a centralised system, whether working alone or as part of a team, with a styleguide document to work from. In this way, I can reuse combinations of html elements to build up `.gl-Navigation`, `.gl-Sidebars`, `.gl-PullQuote`s, image groups, `.gl-Box`es, `.gl-Button`s, and any other components one might like to see peppered across your website or app.

I personally take the [Brutalist](#) approach to user-interfaces, with a “You aren’t going to need it” (YAGNI) mentality. So many designs out there are pure fluff: nice to look at but essentially overkill. I’ve worked with tiny frameworks and others as large as Bootstrap, and let me tell you, CSS is a pain to work with at scale, whereas simple frameworks tend to be a joy to work with (check out [Pico CSS](https://picocss.com/)).

On the flip side of this, users have become accustomed to slick customer journeys, design, animations, or motion, so it really depends on your perspective. I’ve worked in tech long enough to know that good design lasts, and the digital death spiral gives you around 2-3 years before the [internet rots](#). All but one of my projects is dead. Long gone. Bye bye. Never to return. It’s a sad fact of life that a digital “body of work” (websites) is doomed from the offset. You thought cars depreciated in value? Just look at the junk yard of CSS!

Don’t sweat the small stuff; focus on more important things.

### Don’t edit Global styles with a Page or Section

> **TL;DR:** `gl`obal means it's the same site-wide, full stop.

NEVER style a `.gl-` element with a `#page` or `#section`. — always use a `-variant` class on the `.gl-` parent. A `-variant` extension should be similar to, but different from it’s cousin; any more than a handful of differences and you should be reaching for another `.gl` class completely, or a unique `#section` — not a variant class.

### Grid-level elements

> Grid-level elements can also be globals.

For an example, my Anki Flashcards has a `#demo` page that’s replicated a couple of times for different `.gl-Card` types. This could very well be a `.gl-Demo` wrapper with `.gl-DemoGrid` and `.gl-DemoGrid .column-two` and so on. Alternatively, you could make every single page unique, with their own grid styles. Possibly better to not repeat yourself though.

### Here’s a good example of Global styles

```html
<!-- Start with the simplest thing possible
     ======================================
     Start with basic HTML and your specimen file. Only then move onto 
     global styles. Global, Page, and Section CSS can be moved up and
     down the hierarchy. As your app develops, you'll see patterns 
     emerging, so change where it seems appropriate.

     TL;DR: Only add `.gl` design-system CSS where essential -->

<main id="home">
  <nav class="gl-Nav">
    <ul> <!-- plain HTML is fine here -->
	  <li>Navigation item</li>
	  <!-- You could probably just use a `<i>` tag for this -->
	  <li class="gl-Nav_Icon">Navigation item with icon</li>
    </ul>
    <footer>
	  <div class="gl-Nav_SignUp"> <!-- A handy wrapper for flexbox -->
	    <form action="post">      <!-- This is in our specimen file -->
	  	  <label>Signup for our Newsletter</label>
	      <input class="gl-Nav_Input" type="email">
          <!-- Buttons are generally standalone styles -->
		  <button class="gl-Button-blue">Join us!</button>
	    </form>
	  </div>
    </footer>
  </nav>
</main>
```

```css
.gl-Nav ul li  { .. } /* Use plain HTML tags where we can             */
.gl-Nav_Icon   { .. } /* Use ECSS style to make globals stand out     */
.gl-Nav_SignUp { .. } /* `CamelCase` naming with specific purposes    */
```

### Similar but not the same

> What happens if our `.gl-Nav` needs slightly changing for a specific page? How do we handle that situation cleanly?

We could use a `.gl-Nav-variant` like `-about` for a variant of our navigation (on our `#about` page. You should limit the amount of things you change here. If it's more than a handful of styles, it should probably be a   `#unique` section that lives inside the `#about` page.

```css
.gl-Nav {
  /* `&` only works with `.less` files, not plain CSS */
  &-about {
    ul li { .. }
	.gl-NavIcon { .. }
  }
}
```

Take care when using variants though. Ask yourself the question, “is it really necessary to create this variant?” do you _really_ need something other than plain old `.gl-Nav`? You could probably do without. Perhaps there’s a documentation page and you need more information than you’ve got in other pages. Fine, but why not just create a `.gl-Documentation` child and have it live inside the `.gl-Nav`?

It could be that what you’re really looking for is a `#page` specific style, not a global one.

## Pages

> A Page name should be as short as possible, ideally a single word such as `#home`.
> A page is anything that lives as a single view (or page). It’s a unique part of the app.

A page is unique website `#blog`, `#admin`, `#about`, or other view that is distinct and only visible ONCE and in one place. You can _maybe_ cheat, for instance if a `#blog` has a landing page and individual page, it’s probably OK to use the same ID for both.

### Repeatable styles, unique to the Page

> **We use `.snake-case` for our Page styles.**
> Small groups of repeating elements or singletons unique to a page go here too.
> It’s unlikely there’ll be too many of these. If it’s repeatable, it’s often a Global style.

Perhaps you have some data-vis on one page, or a bunch of posts somewhere. It’s a fine-line between what’s in your design-system (globals), and what’s unique to a page, but if it’s only visible in one view, it’s probably a `#page .style`.

These styles will live only within a specific page, but they’re not unique (like a `#section`). They’ll likely be on the page in a few places — perhaps you’ve got some `.img-boxes` or a `.list-item` or two. These aren’t part of the design-system, but **should be viewable somewhere**, probably as template files so you can easily view how they’re put together. That should be obvious.

### If plain HTML tags are descriptive, use them!

> Elements like `header` and `section` are descriptive enough to live without a class attached to them. HTML like `p` and `small` aren’t so descriptive, so you might want to  give these proper class names.

Where unique styling is needed, consider using descriptive HTML5 parent elements, or a simple `.wrapper` class, and plain HTML tags to style. If a `p` or `ul` isn't descriptive enough, then, and ONLY then, start to add unique `.classes` and `#sections`.

### An example of our Page-level styles

```html
<!-- Singletons or small groups of REPEATING HTML
     --------------------------------------------
     Again, start with the simplest thing possible. Here there's no
     need to create complex nested classes. They're singular or small
     groups of items, so they can be given a generic class, unique to
     that page. These would be migrated to Globals if they need to be used 
     in a different area of the site. -->

<main id="admin">
  <section> <!-- Remember, RAW Html is the ideal! -->
    <header>
	  <h3>Dashboard heading</h3> <!-- Same as `specimen.html` -->
      <h4 class="subheading">Time to play!</h3> <!-- A generic class -->
    </header>
    <div class="controls">
      <form>
		<div class="checkboxes">
		 <input class="checkbox">
      	 <input class="checkbox">
      	 <input class="checkbox">
        </div>
      	<div class="sliders">
		  <!-- Possibly a bad example here:
           Why would we have a `.gl-` element inside a `.sliders`
           parent and why wouldn't we just have a raw styled slider? -->
		  <input class="gl-Slider-admin">
		  <p class="slider-text">A label for the slider</p>
          <input class="gl-Slider-admin">
		  <p class="slider-text">A label for the slider</p>
        </div> <!-- /end .sliders -->
    </div> <!-- /end .controls -->
</main>
```

```css
/* We scope everything to the `#page` ID in this instance. We aim to keep
styles self-contained so they can be moved around, but this isn't always
possible (or desirable). We also have a global `-variable` class here,
named after the page */

.gl-Slider-admin { .. } /* Our variable design-system style */
/* ~~#admin .gl-Slider { .. } NEVER style a global element like this!~~ */

#admin section header        { .. } /* Plain old HTML is fine here  */
#admin .subheading           { .. } /* Might not be specific enough */
#admin .checkboxes .checkbox { .. }

/* Where we can add simplicity, do so
/* ----------------------------------
/* Look at our class below. This is a good candidate to move into our
`specimen.html` or `.gl-`obal styles */

#admin .slider-text    { .. }

/* Move to the specimen file (it becomes position dependant) */
form input[type="slider"] + label { .. }

/* Or, move to a global style that's a group of elements     */
.gl-Slider      { .. }
.gl-SliderLabel { .. }
```

### Remember — don’t edit Global styles inside a `#page`!

A helpful reminder. Global styles that change (slightly) within another page should be given a `.gl-Style-variable` class based on the page name, it’s colour, or whatever. It means that it’s a self-contained style that isn’t bound to the `#page` ID but can be moved anywhere. Below we have an example of a self-contained `.gl-Slider-admin` input.

```html
<main id="admin">
  <div class="sliders">
    <input class="gl-Slider-admin">
    <p class="slider-text">A label for the slider</p>
    <input class="gl-Slider-admin">
    <p class="slider-text">A label for the slider</p>
  </div> <!-- /end .sliders -->
</main>
```

Has it become very different to its global cousin? It might be time to make it a new `.gl-` parent class (if we’re using on multiple pages), a unique `#section`, or repeating `#page .class`.

----

## Sections

> Explicit is better than implicit.
> GPS says a Section should always be nested inside a Page …
> But with Thema that’s not always necessary!

What happens when we have unique chunks of code in our Page? How do we name large chunks of heavily nested CSS? GPS advocates for nesting our Sections inside the Page, like so: `#admin #dashboard`. It also says to nest children classes inside the Section. To me, this leads to very verbose code, or code that isn’t quite descriptive enough.

```css
#admin #dashboard .visuals .controls .settings .search .box { .. }
#admin #dashboard .box { .. }
```

Perhaps it’s just habit, or aesthetic, but I’m don’t like when these grow long.

### What can we do instead?

> Enter “Thema Styles” 

We start off by using GPS style, keeping `#page` and `#section`  names descriptive and short (one word ideally). Our other `.class`  names use `.snake-case-style`. We then switch to use what I call "Thema styles". It's based on [Thema codes](#) you get in books.

### Thema styles

> **⚠️ This is a work-in-progress. It could be great, or it good be a dumb thing to do.**
> Thema codes work with a parent code and nested children codes. You can “chop off” parts of the code from the right, and its parent codes are still visible. 

Allows for a very flat style of writing code (or at least, you only need to nest the thing you're actually styling). It's straightforward to reason about — it’s obvious where the code should live, and which element is it’s parent — and looks better to my eyes (for now at least!).

#### The downsides of "Thema Styles"

> It’s pretty low risk, but with MASSIVE sites, still a risk.

The only issue with "Thema style" is that styles are dependant on position, so if they're moved, they need re-naming! This is contrary to `.gl-`obal styles where we prefer to keep things self-contained and flexible to move. It’s also _possible_ (but not likely) that our more terse classes will come up somewhere else in our code base.

GPS also tries to rid of us of ugly BEM-style CSS, which, depending on your perspective, Thema styles does or doesn’t achieve!

#### Why should I bother with this?

A more generic style of writing could work just fine. Especially for singletons or elements with only a couple of children, but it's not quite as  descriptive and will possibly lead to longer nested code. Can we write `.dvcs-search-box` -> `#home .search-box`? Probably. But 
if we want more specificity, it'd become `#home #dashboard .visual-controls .search-box` and so on. It grows quickly.

**There's no reason we need that level of specificity!** It's arduous and superfluous to write that way. 3 levels deep is enough, but if we do that in a generic style, it must be OBVIOUS, with great `/* comments */`  and context (it's surrounding styles), so we know that this class is a child of another, in a certain order, and so on.

### How Thema Styles look

```html
<!-- Thema Styles and "Chopping Off" the letters
     -------------------------------------------
     Here I'm not following GPS religiously. I'm scoping the child
     elements of `#dashboard` with `.dash-`. I could've just as easily
     used `#dashboard .vcs-search-box` instead. -->

<main id="admin">
  <section id="dashboard">
    ...
	<div class="dash-visuals"> <!-- Scoped to `#section` name -->
      <!-- You "chop off" the first letter of it's parent -->
	  <aside class="dash-v-controls"> <!-- Chop off to get `-visuals` -->
        <form class="dash-vc-settings"> <!-- .v. `-controls-` -->
		  <div class="dash-vcs-wrapper"> <!-- .vc. `-settings` --> 
    	  	<label class="dash-vcs-search"> <!-- .vcs. `-search` -->
			<div class="dash-vcs-search-box"> <!-- and so on -->
			  <label>Search for data</label>
			  <input type="search">
			</div>
		  </div>
        </form>   <!-- /end .dash-vc-settings -->
      </aside>    <!-- /end .dash-v-controls -->
    </div>        <!-- /end .dash-visuals -->
  </section>      <!-- /end #dashboard -->
</main>
```

```css
/* Explicit is better than implicit
/* --------------------------------
/* This GPS style is too verbose, or isn't detailed enough.
/* Content should be grouped with comment headers and context */

#admin #dashboard .visuals .controls .settings .search .box { .. }
#admin #dashboard .box { .. }

/* -------------------------------------------------------------------------
/* Dashboard Visuals
/* ------------------------------------------------------------------------- 
/* With Thema Styles, it's always obvious it's a child of some other
/* element. It's also very clear how many levels deep it is. It's terse,
/* it's easy to follow, it's EXPLICIT. */

#admin #dashboard .vcs-search-box { .. }

#admin .dash-vcs-search-box { .. } /* Alternatively written like ...   */
#admin .dvcs-search-box     { .. } /* Or even more condensed ...        */
```

### How about ZERO nesting?

> Take a walk on the wild side. We need some kind of `.scope` to make sure our styles don’t affect other parts of our site (which for small sites is unlikely). For big teams and big sites, this is definitely something to plan against.

Remember, GPS wants you to nest everything in `#page #section` and for good reason. But if our styles are _descriptive_ enough, we may need no nesting at all! The more explicit the name, the easier flat-nesting becomes:

- You don't _always_ need too many levels of nesting,
- Sometimes creating a new (nested) parent `.scope` is preferable.

#### Reducing levels of nesting

However you manage this, keep your styles nicely grouped, well organised, _consistent_, and obvious to their order, parents, and siblings. It’d also be very helpful to somewhere have a visual of your blocks of elements, be it a template file, a design, or somewhere near your main design-system.

```css
.dash-visuals        { .. } /* Chop off `-v` to get `#dashboard`        */
.dash-v-controls     { .. } /* Chop off `-controls` to get `visuals`    */
.dash-vc-settings    { .. } /* Chop off to get `v-controls`             */
.dash-vcs-search     { .. } /* Chop off to get `vc-settings`            */
.dash-vcs-search-box { .. } /* No need to compress this down further    */

/* Use HTML if you can, the following is perfectly valid */
.dash-vcs-search-box input[type="search"] { .. }
```

#### Creating a new (nested) parent `.scope`

For fewer nested elements, you might want to try converting one of the child elements into its own parent. So rather than `.dash-vcs-search-box` it would become:

```css
#dashboard .dash-visuals  { .. } /* Parent containing `.dash-search` */

#dashboard .dash-search   { .. } /* Distinct child of Parent.        */
#dashboard .dash-s-box    { .. } /* Everything inside `.dash-search` */
#dashboard .dash-sb input { .. } /* Keeping things short and tidy    */
```

In the above example, it’s unnecessary to use Thema Styles in `.dash-sb` where you could simply use `.dash-search-box`. If you _really_ wanted to condense it down again, you could simply write:

```css
/* Here we use Thema Styles */
#admin #dashboard .visuals     { .. }
#admin #dashboard .v-controls  { .. }
#admin #dashboard .vc-settings { .. }
#admin #dashboard .vcs-wrapper { .. }

/* We use generic classes, but it's actually a child of `.vcs-wrapper`
/* For smaller, less nested elements, this might be nicer than Thema */
#admin #dashboard .search      { .. }
#admin #dashboard .search-box  { .. }
#admin #dashboard .search-box input[type="search"] { .. }
```

### Fighting against scope creep 

So as we’ve seen, you could use Thema style, GPS styles, or both! If you’re one of the risk averse, you can always wrap everything in a `#page #section` to avoid scope leaking. I think, in general Thema style with a scoped “short code” should be absolutely fine.

----
### Naming classes: How long is too long? 

I’d agree that ECSS style `.scope-ParentElement_ChildElement` classes are very long. I feel we can keep the best of it (scoped to its parent or section, CamelCase) and ditch the overly long style. A couple of examples might be:

```html
<aside class="about-PullQuote"> <!-- Scope our class to the page -->
  <blockquote></blockquote>     <!-- use raw HTML here -->
  <img class="about-PullQuote" src="" alt=""> <!-- #! remove class -->
  <small class="company"></small>  <!-- context: comments/class -->
  <footer><p>something <b>here</b></p></footer> <!-- 4 levels deep -->
</aside>
```

### Some more notes on nesting

> **Whichever method you pick, be consistent.**
> Avoiding the downsides of ECSS, but keeping the upsides.

GPS plums for a nested style of writing, but ECSS teaches us a flat style. Pick one! I’m struggling to get used to a heavily nested style of writing CSS, as it’s nice to quickly scan the left-hand side of a file and pick out a `.contact-form-error` rather than a deeply nested style. It’s a question of personal taste and style, but I’d go with the following rules:

- `.gl-`obal elements can have nested classes.
- A `#section` should **always** be nested inside a `#page`, somehow.
- **Never nest styles more than 3 levels deep!** (unless short like `p` or `ul`)
- Don’t use too many `#section`s if you can avoid it.
- Avoid overly specific selectors like `>`.

Let’s just quickly go through a few of those rules. Where selectors are short, such as `p` and `b` you can nest more than 3 levels deep, but reduce down as much as possible. Write `#sections` a bit like you might do with Markdown. You’d probably only have one `# h1` and a handful of `## h2` within your blog post.

Getting too specific with `>` means if you add a wrapper, or move your code around, it’s going to break. GPS isn’t very accommodating to moving things around, but try to isolate your `.child` code enough so you can move a `.button` from A to B without it breaking the styling — be aware you may have to refactor your code if you move elements around.

Nesting  `-hover` or `-action` states can be handy, using `&:hover`, but you could just as well add them as separate styles. Where `@media` queries are needed, I generally prefer to keep them close to the element I’m styling, even if this results in many `@media` styles. I also nest them, but this is optional:

```css
.gl-Wrapper { @media screen and (min-width: 400px) { .. } }
```

The above will only work if you’re using a compiler like [Less](https://lesscss.org/) or [Sass](https://sass-lang.com/) (which I only use very lightly). For regular CSS, the `@media` rules could live with, but outside of, the main class of your element:

```css
.gl-Wrapper { .. }

@media screen and (min-width: 400px) {
	.gl-Wrapper { .. }
}
```

### Keeping similar styles together

A helpful tip learned from ECSS is to keep component files together, so you could have a `/page` folder with `page.css` and `page.js` if you so desire. Same for `.gl-` components if they have a `.js` or `.mustache` (or other template) files.

### Using `#fragments` for links

So long as your `#fragment` isn’t used for styling, it can be moved around pretty much anywhere. Just make it obvious that it’s used for linking or targeting for javascript etc. For example, you could name your links `#link-fragment` to avoid confusion.

[^1]: GPS is a [method](https://medium.com/@jescalan/bem-is-terrible-f421495d093a) for naming and isolating CSS. I’m not using GPS religiously and am trying out my own “Thema Style” CSS, so this article might be a little convoluted right now. I’m basically using ECSS conventions for globals, GPS for the base-level isolation, but introducing `.thema-styles` and `.thema-s-children` so I can flatten the CSS, which I personally prefer. `#page #section` with two IDs still feels weird to me (and looks weird in the editor!).

[^2]: This is where I feel GPS gets a bit wishy washy. _“All page-specific styles should be scoped under this `#page` ID”_ and _“Any CSS that you write that is not global should always be scoped under a page or view’s ID”_. A `#section` is a unique bit of the page, but if _“later on you notice that the same style is actually used elsewhere on the same page, you can pull it up to a page-specific style”_, and then on up the chain until you hit `global` elements, site-wide. This is heavily leaning towards _nested_ CSS and I’m not sure that I like that.
